packages/lib-classifier/src/plugins/drawingTools/experimental/models/marks/FreehandLine/FreehandLine.js

```diff
@@ -3,19 +3,37 @@ import { FreehandLine as FreehandLineComponent } from '@plugins/drawingTools/com
 import { Mark } from '@plugins/drawingTools/models/marks'
 import { FreehandLineTool } from '@plugins/drawingTools/models/tools'
 import { FixedNumber } from '@plugins/drawingTools/types/'
+import { action, toJS } from 'mobx'
 
 const LINE_RESOLUTION = 0.5
 const MINIMUM_POINTS = 20
+const CLOSE_DISTANCE = 10
 
 const SingleCoord = types.model('SingleCoord', {
   x: FixedNumber,
   y: FixedNumber
 })
 
+const actions = [];
+
+const ActionCoor = types.model('ActionCoor', {
+  0: types.number,
+  1: types.number
+});
+
+const ActionType = types.model('ActionType', {
+  type: types.string,
+  points: types.array(ActionCoor),
+  pointIndex: types.number,
+  isReversed: types.boolean,
+});
+
 const FreehandLineModel = types
   .model('FreehandLineModel', {
     pathX: types.array(FixedNumber),
-    pathY: types.array(FixedNumber)
+    pathY: types.array(FixedNumber),
+    drawActions: types.array(ActionType),
+    redoActions: types.array(ActionType),
```

Comment:
MobX State Tree has middleware for recording a history of patches applied to the model. You can then undo and redo the patch history without having to write your own custom history manager.

Reply:
The mark gets submitted with the classification, so you may also find that classifications now contain `drawActions` and `redoActions`. 

You can use volatile storage for values that aren’t expected to persist in snapshots.

Reply:
The MobX State Tree was not able to manage the state tree for comprehensive undo/redo actions effectively, particularly with splicing a path through its beginning+end. This code was moved to volatile storage so that it isn't persisted in snapshots.